keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,684 @@
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a Kubernetes admin using IPv6, I want to be able to follow the IPv6 address
+planning best practices, being able to assign /64 prefixes to my end subnets.
+
+#### Story 2
+
+As a Kubernetes user I want to be able to select any IP address from the range I
+have configured.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The service IP allocator uses a bitmap to store and track the assigned IP.
+This bitmap is snapshoted in etcd, so multiple apiserver, can assign IP
+addresses without conflicts.
+To avoid performance issues because of the size of the bitmap, the apiserver limits
+both the size of the bitmap in IPv6 and the max subnet size for IPv4 and IPv6.
+However, we run into a incongruency here, because the IPv4 cidr prefix length
+is limited to /12, but not the bitmap size, so we can have a bitmap of 2^20 bits.
+For an IPv6 we limit the bitmap size to 2^16 bits, so despite allowing a /112 we only
+allocate from the low 16 bits of the subnet.
+
+A bitmap of 2^64 is not realistic, mainly if we have to serialize and deserialize it
+continuously to snapshot/recover it in/from etcd.
+
+The solution proposed consist in:
+
+- limit the maximum bitmap size to
+
+```
+<<[UNRESOLVED max bitmap size ]>>
+
+1. 2^20 bits = 1048576
+2. 2^16 bits = 65536
+
+<<[/UNRESOLVED]>>
+```
+
+- limit the maximum cluster cidr prefix length
+
+```
+<<[UNRESOLVED apiserver subnet validation ]>>
+In my personal preferred order
+
+1. service cluster-cidr allows max prefix length IPv4 /12 and IPv6 /64 (and fewer)
+2. service cluster-cidr allows max prefix length IPv4 /12 and IPv6 only /64 or fewer that /112
+2. service cluster-cidr allows max prefix length IPv4 /12 and IPv6 only /64 (not fewer)
+3. service cluster-cidr allows any prefix
+
+<<[/UNRESOLVED]>>
+```
+
+- If the CIDR is larger than that fixed size, set the bit (IP number) % (bit array size).
```

Comment:
This feels like a lot of work to introduce a new failure mode.  What's really motivating this change?  This sort of behavior fails the "least surprise" test, IMO.  

If there's REALLY a need to be able to allocate from anywhere in the whole /64 space, maybe we should focus on a better representation than a dumb bitmap?

Reply:
This is a spinoff of https://github.com/kubernetes/kubernetes/pull/90115 but I keep being unhappy with every alternative we come up with.

The motivation is that forcing people to use a non-/64 subnet length just feels wrong. It says "Yes, fellow "hackers", we support the Internet Protocol Version Six for sending your Transmission Control Protocol packages over". It's not entirely wrong, but it gives off strong "we don't really understand what we're talking about" vibes.

So the goal is to get rid of that limitation, without causing any other problems, but the "not causing any other problems" part keeps screwing things up.

(I am not sure I like this approach better than any of the previous suggestions.)

Reply:
> This feels like a lot of work to introduce a new failure mode. What's really motivating this change? This sort of behavior fails the "least surprise" test, IMO.

in my personal opinion /64 should be the only allowed prefix for the service subnet, there is a lot of literature about that, but that will be the "least surprise" test, so I think that at least we should allow users to use /64 for the service subnet

> If there's REALLY a need to be able to allocate from anywhere in the whole /64 space, 

IPv6 protocol was designed to be autoconfiguring, personally, I don't think there should be need to assign addresses manually from the whole range. However, this really breaks current behavior where users can assign from the whole range.

> maybe we should focus on a better representation than a dumb bitmap?

I've tried that with 2 different data structures, added in the alternatives sections, did it with roaring bitmaps and a map[int64]struct{} , it seems too disruptive and too much effort for something that as Dan said 

> So the goal is to get rid of that limitation, without causing any other problems, but the "not causing any other problems" > part keeps screwing things up.